### PR TITLE
[SPARK-18375][SPARK-18383][BUILD][CORE]Upgrade netty to 4.0.42.Final

### PR DESCRIPTION
**Pull request by witgo**
## What changes were proposed in this pull request?

One of the important changes for 4.0.42.Final is "Support any FileRegion implementation when using epoll transport netty/netty#5825".
In 4.0.42.Final, `MessageWithHeader` can work properly when `spark.[shuffle|rpc].io.mode` is set to epoll

## How was this patch tested?

Existing tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -2222,6 +2223,9 @@ private[spark] object Utils extends Logging {
         isBindCollision(e.getCause)
       case e: MultiException =>
         e.getThrowables.asScala.exists(isBindCollision)
+      case e: NativeIoException =>
+        e.getMessage().startsWith("bind() failed: ") ||
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Seems OK. Do you need to check for a null message though, like above? In fact, the condition above for `BindException` could be `e.getMessage != null || isBindCollision(e.getCause)`

Comment (witgo) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: From the netty code, the return of `NativeIoException.getMessage` is not null.
The only way to  create a NativeIoException instance is to call [Errors.newIOException](https://github.com/netty/netty/blob/4.0/transport-native-epoll/src/main/java/io/netty/channel/unix/Errors.java#L116)

Comment (witgo) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: OK, I will modify the code.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master/2.1

**Comment by witgo** [comment]
@srowen thank you

**Pull request closed without merging**